flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/BlockStatementSplitter.java

```diff
@@ -0,0 +1,417 @@
+ *     b = a * 2;
+ *     c = b * 2;
+ *     System.out.println(b);
+ * </code></pre>
+ *
+ * <pre><code>
+ * myFun_whileBody0_0_ifBody1(int a) ->
+ *     b = a * 3;
+ *     System.out.println(b);
+ * </code></pre>
+ */
+@Internal
+public class BlockStatementSplitter {
+
+    private final String code;
+
+    private final List<BlockStatementSplitter> children = new ArrayList<>();
+
+    private final String parameters;
+
+    private BlockStatementVisitor visitor;
+
+    /**
+     * Initialize new BlockStatementSplitter.
+     *
+     * @param code a code block that should be rewritten.
+     * @param parameters parameters definition that should be used for extracted methods.
+     */
+    public BlockStatementSplitter(String code, String parameters) {
+        this.code = code;
+        this.parameters = parameters;
+    }
+
+    /**
+     * This method extracts statements from IFs, ELSE's and WHILE blocks from block code used during
+     * initialization of this object. Every entry of returned map can be seen as new method nam (map
+     * key) and method's body. The block names will be prefixed with provided context.
+     *
+     * @param context prefix for extracted blocks.
+     * @return a map of block name to block statements mappings. The key can be interpreted as name
+     *     of extracted block/method and corresponding List represents individual statements (block'
+     *     lines) for this block.
+     */
+    public Map<String, List<String>> extractBlocks(String context) {
+
+        this.visitor = new BlockStatementVisitor(code, context, parameters);
+        JavaParser javaParser = new JavaParser(visitor.tokenStream);
+        javaParser.getInterpreter().setPredictionMode(PredictionMode.SLL);
+        visitor.visitStatement(javaParser.statement());
+
+        Map<String, List<String>> extractedBlocks = new HashMap<>();
+        for (BlockStatementVisitor child : visitor.children) {
+            int counter = 0;
+            for (ContextTextPair extractedBlock : child.extractedSingleBlocks) {
+                ParserRuleContext parserRuleContext = extractedBlock.parserRuleContext;
+                if (parserRuleContext instanceof BlockStatementContext) {
+                    StatementContext statement =
+                            ((BlockStatementContext) parserRuleContext).statement();
+
+                    if (statement != null
+                            && (statement.IF() != null
+                                    || statement.ELSE() != null
+                                    || statement.WHILE() != null)) {
+
+                        BlockStatementSplitter splitter =
+                                new BlockStatementSplitter(
+                                        extractedBlock.ruleText, this.parameters);
+                        Map<String, List<String>> rewrite =
+                                splitter.extractBlocks(child.context + "_" + counter++);
+                        this.children.add(splitter);
```

Comment:
You use recursions both in `BlockStatementSplitter#extractBlocks` and in `BlockStatementVisitor#visitStatement`. This doesn't look right to me. Could you explain your logic in detail?